File: C2/PDOO/Practicas/DeepSpaceRuby/lib/Weapon.rb

#@author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio

#encoding:utf-8

require_relative "WeaponToUI"

require_relative "WeaponType"

module Deepspace
  class Weapon
    
    attr_reader :type, :uses, :name
    
    def initialize(n, t, u)
      @name = n
      @type = t
      @uses = u
    end
   
    def self.newCopy(s)
      new(s.name,s.type, s.uses)
    end
    
    def getUIversion
      return WeaponToUI.new(self)
    end
    
    def power()
      return @type.power
    end
    
    def useIt()
     if @uses > 0
       @uses = @uses - 1
       return @type.power
     else
       return 1.0
     end
    end
    
    def to_s
      cad= "Nombre: #{@name}, Usos: #{@uses}"
      if(@type==WeaponType::MISSILE)
        cad=cad+", Tipo: MISSILE\n"
      elsif(@type==WeaponType::PLASMA)
        cad=cad+", Tipo: PLASMA\n"
      else
        cad=cad+", Tipo: LASER\n"
      end
      return cad
    end
    
  end
end

File: C2/PDOO/Practicas/DeepSpaceJava/src/deepspace/Weapon.java
package deepspace;



/**
 *
 * @author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio
 */
public class Weapon implements Copyable <Weapon>, CombatElement {
    private String name;
    private WeaponType type;
    private int uses;
    
    public Weapon(String new_name, WeaponType new_type, int new_uses ){
        name=new_name;
        type=new_type;
        uses=new_uses;
    }
    
    public Weapon(Weapon s){
        name=s.name;
        type=s.type;
        uses=s.uses;
    }
    
    public WeaponType getType(){
        return type;
    }
    
    @Override
    public int getUses(){
        return uses;
    }
    
    public float power(){
        return type.getPower();
    }
    
    WeaponToUI getUIversion(){
        return new WeaponToUI(this);
    }
    
    @Override
    public float useIt(){
        if(uses>0){
            uses--;
            return power();
        }
        else{
            return 1.0f;
        }
    }
    
    WeaponToUI getUIVersion(){
        return new WeaponToUI(this);
    }
    
    public Weapon copy(){
        return new Weapon(this);
    }
    
     public String toString(){
        return "Nombre: "+this.name+" Tipo: "+this.getType()+"("+power()+") Usos: "+this.getUses()+"  ";
    }
}
